File: clients/ios/Lextures/Core/LMS/BehaviorLogic.swift
import Foundation

enum BehaviorAwardMode: String, CaseIterable {
    case award
    case referral
}

struct HallPassCountdown: Equatable {
    var remainingSeconds: Int
    var isExpired: Bool
    var isOverdue: Bool
}

enum BehaviorLogic {
    static let studentRoles: Set<String> = ["student", "learner"]
    static let hallPassDestinations = ["bathroom", "office", "library", "nurse", "other"]
    static let defaultPassMinutes = 5

    static func studentRoster(from enrollments: [CourseEnrollment]) -> [CourseEnrollment] {
        enrollments.filter { studentRoles.contains($0.role.lowercased()) }
    }

    static func studentLabel(_ enrollment: CourseEnrollment) -> String {
        let name = enrollment.displayName?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        if !name.isEmpty { return name }
        return "Student"
    }

    static func activeCategories(_ categories: [BehaviorCategory]) -> [BehaviorCategory] {
        categories.filter(\.active)
    }

    static func positiveCategories(_ categories: [BehaviorCategory]) -> [BehaviorCategory] {
        activeCategories(categories).filter(\.isPositive)
    }

    static func negativeCategories(_ categories: [BehaviorCategory]) -> [BehaviorCategory] {
        activeCategories(categories).filter(\.isNegative)
    }

    static func awardPayload(
        studentIds: Set<String>,
        categoryId: String,
        note: String?
    ) -> [PBISAwardInput] {
        let trimmedNote = note?.trimmingCharacters(in: .whitespacesAndNewlines)
        return studentIds.map { studentId in
            PBISAwardInput(
                studentId: studentId,
                categoryId: categoryId,
                points: 1,
                note: trimmedNote?.isEmpty == false ? trimmedNote : nil
            )
        }
    }

    static func isActiveHallPass(_ pass: HallPass) -> Bool {
        let status = pass.status.lowercased()
        return status == "requested" || status == "approved"
    }

    static func hallPassCountdown(
        pass: HallPass,
        now: Date = Date()
    ) -> HallPassCountdown? {
        guard pass.status.lowercased() == "approved" else { return nil }
        guard let approvedAt = parseDate(pass.approvedAt) else { return nil }
        let estimated = pass.estimatedMins ?? defaultPassMinutes
        let deadline = approvedAt.addingTimeInterval(TimeInterval(estimated * 60))
        let remaining = Int(deadline.timeIntervalSince(now))
        let overdue = pass.overdue == true || remaining <= 0
        return HallPassCountdown(
            remainingSeconds: max(remaining, 0),
            isExpired: remaining <= 0,
            isOverdue: overdue
        )
    }

    static func formatCountdown(_ countdown: HallPassCountdown) -> String {
        let total = countdown.remainingSeconds
        let minutes = total / 60
        let seconds = total % 60
        return String(format: "%d:%02d", minutes, seconds)
    }

    static func destinationLabel(_ destination: String) -> String {
        switch destination.lowercased() {
        case "bathroom": return L.text("mobile.hallpass.destination.bathroom")
        case "office": return L.text("mobile.hallpass.destination.office")
        case "library": return L.text("mobile.hallpass.destination.library")
        case "nurse": return L.text("mobile.hallpass.destination.nurse")
        default: return L.text("mobile.hallpass.destination.other")
        }
    }

    static func statusLabel(_ status: String) -> String {
        switch status.lowercased() {
        case "requested": return L.text("mobile.hallpass.status.requested")
        case "approved": return L.text("mobile.hallpass.status.approved")
        case "denied": return L.text("mobile.hallpass.status.denied")
        case "returned": return L.text("mobile.hallpass.status.returned")
        default: return status
        }
    }

    static func storedPassKey(sectionId: String) -> String {
        "hallPass:\(sectionId)"
    }

    private static func parseDate(_ value: String?) -> Date? {
        guard let value, !value.isEmpty else { return nil }
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let date = formatter.date(from: value) { return date }
        formatter.formatOptions = [.withInternetDateTime]
        return formatter.date(from: value)
    }
}

File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/BehaviorLogic.kt
package com.lextures.android.core.lms

import java.time.Instant
import java.time.format.DateTimeFormatterBuilder
import java.time.temporal.ChronoUnit

enum class BehaviorAwardMode {
    Award,
    Referral,
}

data class HallPassCountdown(
    val remainingSeconds: Int,
    val isExpired: Boolean,
    val isOverdue: Boolean,
)

object BehaviorLogic {
    private val studentRoles = setOf("student", "learner")
    val hallPassDestinations = listOf("bathroom", "office", "library", "nurse", "other")
    const val defaultPassMinutes = 5

    fun studentRoster(enrollments: List<CourseEnrollment>): List<CourseEnrollment> =
        enrollments.filter { it.role.lowercase() in studentRoles }

    fun studentLabel(enrollment: CourseEnrollment): String {
        val name = enrollment.displayName?.trim().orEmpty()
        return name.ifEmpty { "Student" }
    }

    fun activeCategories(categories: List<BehaviorCategory>): List<BehaviorCategory> =
        categories.filter { it.active }

    fun positiveCategories(categories: List<BehaviorCategory>): List<BehaviorCategory> =
        activeCategories(categories).filter { it.isPositive }

    fun negativeCategories(categories: List<BehaviorCategory>): List<BehaviorCategory> =
        activeCategories(categories).filter { it.isNegative }

    fun awardPayload(
        studentIds: Set<String>,
        categoryId: String,
        note: String?,
    ): List<PbisAwardInput> {
        val trimmed = note?.trim().orEmpty()
        return studentIds.map { studentId ->
            PbisAwardInput(
                studentId = studentId,
                categoryId = categoryId,
                points = 1,
                note = trimmed.takeIf { it.isNotEmpty() },
            )
        }
    }

    fun isActiveHallPass(pass: HallPass): Boolean {
        val status = pass.status.lowercase()
        return status == "requested" || status == "approved"
    }

    fun hallPassCountdown(pass: HallPass, now: Instant = Instant.now()): HallPassCountdown? {
        if (pass.status.lowercase() != "approved") return null
        val approvedAt = parseInstant(pass.approvedAt) ?: return null
        val estimated = pass.estimatedMins ?: defaultPassMinutes
        val deadline = approvedAt.plus(estimated.toLong(), ChronoUnit.MINUTES)
        val remaining = ChronoUnit.SECONDS.between(now, deadline).toInt()
        val overdue = pass.overdue == true || remaining <= 0
        return HallPassCountdown(
            remainingSeconds = maxOf(remaining, 0),
            isExpired = remaining <= 0,
            isOverdue = overdue,
        )
    }

    fun formatCountdown(countdown: HallPassCountdown): String {
        val minutes = countdown.remainingSeconds / 60
        val seconds = countdown.remainingSeconds % 60
        return "%d:%02d".format(minutes, seconds)
    }

    fun destinationLabelRes(destination: String): String = when (destination.lowercase()) {
        "bathroom" -> "mobile_hallpass_destination_bathroom"
        "office" -> "mobile_hallpass_destination_office"
        "library" -> "mobile_hallpass_destination_library"
        "nurse" -> "mobile_hallpass_destination_nurse"
        else -> "mobile_hallpass_destination_other"
    }

    fun statusLabelRes(status: String): String = when (status.lowercase()) {
        "requested" -> "mobile_hallpass_status_requested"
        "approved" -> "mobile_hallpass_status_approved"
        "denied" -> "mobile_hallpass_status_denied"
        "returned" -> "mobile_hallpass_status_returned"
        else -> status
    }

    fun storedPassKey(sectionId: String): String = "hallPass:$sectionId"

    private fun parseInstant(value: String?): Instant? {
        if (value.isNullOrBlank()) return null
        return runCatching {
            Instant.from(
                DateTimeFormatterBuilder()
                    .appendInstant(3)
                    .toFormatter()
                    .parse(value),
            )
        }.getOrNull()
    }
}
